src/Misc.pm

```diff
@@ -4388,8 +4388,16 @@ sub makeShop {
 
 	# Iterate through items to be sold
 	shuffleArray(\@{$shop{items}}) if ($config{'shop_random'} eq "2");
+	my %used_items;
 	for my $sale (@{$shop{items}}) {
-		my $cart_item = $char->cart->getByName($sale->{name});
+		my $cart_item;
+		for my $item (@{$char->cart->getItems}) {
+			if ($item->{name} eq $sale->{name} and !$used_items{$item->{invIndex}}) {
+				$cart_item = $item;
+				$used_items{$item->{invIndex}} = 1;
+				last
```

Comment:
I like early exit style better than wrapping the whole block in an `if`. Just personal preference, please feel free to ignore me. :)

Here's my suggestion:
```
for my $item (@{$char->cart->getItems}) {
    next if $item->{name} ne $sale->{name};
    next if $used_items{$item->{invIndex}};
    $cart_item = $used_items->{$item->{invIndex}} = $item;
    last;
}
```